[HVAC] Implement TC-TSTAT-2.2

hasty opened the pull request:
This provides a Python implementation of TC-TSTAT-2.2 to replace the unused TC-TSTAT-2.2 YAML test.

Fixes #35748

#### Testing

This should restore the setpoint values after the test, allowing for multiple runs.

hasty committed:
Initial generated test

hasty committed:
Fixed and compared against test plan

hasty committed:
Remove unused TC-TSTAT-2.2 yaml

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+
+                # Test Harness Reads it back again to confirm the successful write of MaxHeatSetpointLimit attribute
+                val = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.MaxHeatSetpointLimit)
+                asserts.assert_equal(val, 2900)
```
```suggestion
                asserts.assert_equal(val, AbsMaxHeatSetpointLimitValue-1)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+            await self.write_single_attribute(attribute_value=cluster.Attributes.OccupiedHeatingSetpoint(OccupiedHeatingSetpointValue), endpoint_id=endpoint)
+
+            # Sends SetpointRaise Command Heat Only
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kHeat, amount=-30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kHeat, amount=-30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+            await self.write_single_attribute(attribute_value=cluster.Attributes.OccupiedHeatingSetpoint(OccupiedHeatingSetpointValue), endpoint_id=endpoint)
+
+            # Test Harness Sends SetpointRaise Command Heat Only
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kHeat, amount=30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kHeat, amount=30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+            await self.write_single_attribute(attribute_value=cluster.Attributes.OccupiedCoolingSetpoint(OccupiedCoolingSetpointValue), endpoint_id=endpoint)
+
+            # Test Harness Sends SetpointRaise Command Cool Only
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kCool, amount=30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kCool, amount=30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+
+        if self.pics_guard(hasCoolingFeature):
+            # Test Harness Sends SetpointRaise Command Cool Only
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kCool, amount=-30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kCool, amount=-30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+
+        if self.pics_guard(hasHeatingFeature or hasCoolingFeature):
+            # Test Harness Sends SetpointRaise Command Heat & Cool
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kBoth, amount=-30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kBoth, amount=-30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+
+        if self.pics_guard(hasHeatingFeature or hasCoolingFeature):
+            # Test Harness Sends SetpointRaise Command Heat & Cool
+            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kBoth, amount=30))
```
```suggestion
            await self.send_single_cmd(cmd=Clusters.Objects.Thermostat.Commands.SetpointRaiseLower(mode=Clusters.Objects.Thermostat.Enums.SetpointRaiseLowerModeEnum.kBoth, amount=30), endpoint=endpoint)
```

fessehaeve commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+    def steps_TC_TSTAT_2_2(self) -> list[TestStep]:
+        steps = [
+            TestStep("1", "Commission DUT to TH"),
+            TestStep("2a", "Test Harness Client reads  attribute OccupiedCoolingSetpoint from the DUT"),
```
Can we store all initial attribute values (mainly min and max values) so that we can restore them at the end of the test? I wasn't able to run this test back to back since by the end of the test some of the attribute values don't make sense anymore.

fessehaeve replied:
The exact issue I saw here is that, by the end of the test, the heatingsetpoint is very close to the abs/maxheatingsetpoint and some tests increment fixed values to heatingsetpoint which would be out of range and fail a test.

hasty replied:
This turned out to be harder than you might expect, as the default values for the thermostat cluster violate its own constraints. Once [PR 36023](https://github.com/project-chip/connectedhomeip/pull/36023) is merged, we can update the code to match, and then this will work.

hasty replied:
Yeah, copied and pasted the wrong one. Meant to reference [PR 10585](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10585), which has since been merged. I've updated this test to restore the values of the setpoints at the time the test started.

hasty committed:
Apply suggestions from code review

Co-authored-by: fesseha-eve <88329315+fessehaeve@users.noreply.github.com>

KishokG commented:
@hasty We have to add steps 19 to 22c as per the test plan. Here is the test plan reference: https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/thermostat.adoc#tc-tstat-2-2-setpoint-test-cases-with-server-as-dut

KishokG commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+from chip.clusters import Globals
+from chip.clusters.Types import NullValue
+from chip.interaction_model import InteractionModelError, Status
+from matter_testing_support import MatterBaseTest, TestStep, async_test_body, default_matter_test_main
```
@hasty this should be update to:  `from chip.testing.matter_testing import MatterBaseTest, TestStep, async_test_body, default_matter_test_main`

hasty replied:
Your PR fixed this, thanks!

KishokG commented:
The following PR: https://github.com/hasty/connectedhomeip/pull/2 needs to be merged into Hasty's branch else the changes needs to be updated manually in this PR.

KishokG committed:
Update TC_TSTAT_2_2.py

fixed below mentioned discrepancies on the code:

Step-3c: The test plan expects a Success response, but the script is checking for a ConstraintError. If the value does not violate any conditions, the response should be Success. In such cases, the test will fail.
Step-4c: If the AUTO feature is enabled, the script writes the UnoccupiedCoolingSetpoint attribute with MaxCoolSetpointLimitValue and checks for a ConstraintError response, which is not required.
Step-5c: The script should check the following condition when the AUTO feature is enabled: Min(MaxHeatSetpointLimit, (UnoccupiedCoolingSetpoint - MinSetpointDeadBand)). However, this condition is missing in the script, and it incorrectly checks for a ConstraintError response.
Step-7a: The write and read verification is missing.
Step-9b: The script incorrectly writes the attribute value using AbsMinCoolSetpointLimitValue instead of MinCoolSetpointLimitValue.
Step-9c: A duplicate read check has been removed.

hasty committed:
Merge branch 'master' into granbery/thermostat_setpoint_limits_test

hasty committed:
Merge pull request #3 from KishokG/patch-2

Update TC_TSTAT_2_2.py

hasty commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,778 @@
+
+    def steps_TC_TSTAT_2_2(self) -> list[TestStep]:
+        steps = [
+            TestStep("1", "Commission DUT to TH"),
```
Kishok's PR fixed this.

nivi-apple commented on src/python_testing/TC_TSTAT_2_2.py:
```diff
@@ -0,0 +1,777 @@
+        steps = [
+            TestStep("1", "Commissioning, already done", is_commissioning=True),
+            TestStep("2a", "Test Harness Client reads  attribute OccupiedCoolingSetpoint from the DUT"),
+            TestStep("2b", "Test Harness Client then attempts  Writes OccupiedCoolingSetpoint to value below the MinCoolSetpointLimit"),
```
```suggestion
            TestStep("2b", "Test Harness Client then attempts to write OccupiedCoolingSetpoint to value below the MinCoolSetpointLimit"),
```

nivi-apple replied:
```suggestion
            TestStep("2b", "Test Harness Client then attempts Writes OccupiedCoolingSetpoint to a value below the MinCoolSetpointLimit"),
```

hasty committed:
Python linter fix for TC_TSTAT_2_2

hasty committed:
Merge branch 'master' into granbery/thermostat_setpoint_limits_test

hasty committed:
Update src/python_testing/TC_TSTAT_2_2.py

Co-authored-by: Nivi Sarkar <55898241+nivi-apple@users.noreply.github.com>

hasty committed:
Restore setpoints to original values at end of test

hasty committed:
Merge branch 'master' into granbery/thermostat_setpoint_limits_test

hasty committed:
Use get_endpoint method to find active endpoint

hasty committed:
Update CI Test Arguments to the apparent new style
